c/driver/common/utils.c

```diff
@@ -119,24 +120,41 @@ AdbcStatusCode BatchToArrayStream(struct ArrowArray* values, struct ArrowSchema*
   return ADBC_STATUS_OK;
 }
 
-void StringBuilderInit(struct StringBuilder* builder, size_t initial_size) {
+int StringBuilderInit(struct StringBuilder* builder, size_t initial_size) {
   builder->buffer = (char*)malloc(initial_size);
+  if (builder->buffer == NULL) return -1;
+
   builder->size = 0;
   builder->capacity = initial_size;
+
+  return 0;
 }
-void StringBuilderAppend(struct StringBuilder* builder, const char* value) {
-  size_t length = strlen(value);
-  size_t new_size = builder->size + length;
-  if (new_size > builder->capacity) {
-    size_t new_capacity = builder->size + length - builder->capacity;
-    if (builder->size == 0) new_capacity++;
-
-    builder->buffer = realloc(builder->buffer, new_capacity);
-    builder->capacity = new_capacity;
+int StringBuilderAppend(struct StringBuilder* builder, const char* fmt, ...) {
+  va_list argptr;
+  int bytes_available = builder->capacity - builder->size;
+
+  va_start(argptr, fmt);
+  int n = vsnprintf(builder->buffer + builder->size, bytes_available, fmt, argptr);
+  va_end(argptr);
+
+  if (n < 0) {
+    return -1;
```

Comment:
in general, maybe return one of the errno codes for consistency with the nanoarrow behavior?

Reply:
Makes sense - will give this a look